sample-viewmodel/src/main/java/com/example/molecule/viewmodel/MoleculeViewModel.kt

```diff
@@ -45,6 +56,21 @@ abstract class MoleculeViewModel<Event, Model> : ViewModel() {
     }
   }
 
-  @Composable
-  protected abstract fun models(events: Flow<Event>): Model
```

Comment:
Consider amending this to

```kotlin
   abstract val initialSeed: Model

  @Composable
  protected abstract fun models(seed: Model, events: Flow<Event>): Model
``` 

Then to produce state in the `MoleculeViewModel`:

```kotlin
  private var seed = initialState

  val models: StateFlow<Model> by lazy(LazyThreadSafetyMode.NONE) {
    moleculeFlow(mode = ContextClock) {
      models(seed, events)
    }.onEach {
      seed = it
    }.stateIn(
      scope = scope,
      started = SharingStarted.WhileSubscribed(5.seconds),
      initialValue = seed,
    )
  }
```

With this change, the seed is explicitly needed in state production, and implementers of `MoleculeViewModel` have to provide an initial state.

When the presented `Flow` is collected from again, the last seen seed is provided so it can be restored.

Reply:
A bit confused here, this part no longer is remembering the Presenter, should the burden of doing that then fall on the consumer instead?
Also, doing this
```
abstract val initialState: Model

private var seed: Model = initialState
```
Inside the `abstract class MoleculeViewModel` gives me a warning of "Accessing non-final property initialState in constructor" for the `seed`
<img width="683" alt="image" src="https://github.com/cashapp/molecule/assets/44558292/02c41d53-1200-4f3c-8263-12449b072c2c">
I must be missing something here 👀

Tried this out here https://github.com/StylianosGakis/molecule/pull/1 and it seems like the warning I am getting does in fact make it so that when I run the app it finds `initialState` to be null because it's not initialized quick enough, so `seed` is also null, so the initial value inside the StateFlow is null, crashing the app.

Stacktrace starts like:
```
java.lang.NullPointerException: Parameter specified as non-null is null: method com.example.molecule.viewmodel.UiKt.PupperPicsScreen, parameter model
  at com.example.molecule.viewmodel.UiKt.PupperPicsScreen(Unknown Source:6)
```

Reply:
Ah sorry about that. The warning is right in that the parent class is dependent on the subclass for a value, the proper way to do this is to pass it in the constructor.

If you added an `initialState` argument to the `MoleculeViewmodel` it should take care of the issue. This argument would then behave like the initial state in the `stateIn` operator on a `Flow`. The seed will be initialized to it and the rest should follow hopefully.

Reply:
I do think separating the presenter from the VM is a great pattern, my current suggestion optimizes for simplicity. Extend the VM, produce state.

Otherwise the presenter factory you have is great granted the returned presenter class is stateless. In your impl it's a class and it's not immediately obvious that on each restart a new instance of the class is created and any local state in the class lost. This is implicit in `models(seed, events) ` because there's no enclosing class body.

Reply:
Okay, I've validated the above. I have:

`MoleculeViewModel` definition:

```kotlin
abstract class MoleculeViewModel<Event, Model>(
  initialState: Model,
  started: SharingStarted = SharingStarted.WhileSubscribed(5.seconds),
  presenter: @Composable (seed: Model, events: Flow<Event>) -> Model,
) : ViewModel() {
  ...

  private var seed: Model = initialState

  val models: StateFlow<Model> by lazy(LazyThreadSafetyMode.NONE) {
    moleculeFlow(mode = ContextClock) {
      presenter(seed, events)
    }.onEach {
      seed = it
    }.stateIn(
      scope = scope,
      started = started,
      initialValue = seed,
    )
  }
  ...
}
```

```kotlin
class PupperPicsViewModel(
  // This service would typically be injected
  service: PupperPicsService = PupperPicsService(),
) : MoleculeViewModel<Event, Model>(
  initialState = Model(
    loading = false,
    breeds = emptyList(),
    currentBreed = null,
    currentUrl = null,
  ),
  presenter = { seed, events ->
    PupperPicsPresenter(
      seed = seed,
      events = events,
      service = service,
    )
  },
)
```

Then the `PupperPicsPresenter`:

```kotlin
@Composable
fun PupperPicsPresenter(
  seed: Model,
  events: Flow<Event>,
  service: PupperPicsService,
): Model {
  var breeds: List<String> by remember { mutableStateOf(seed.breeds) }
  var currentBreed: String? by remember { mutableStateOf(seed.currentBreed) }
  var currentUrl: String? by remember { mutableStateOf(seed.currentUrl) }
  var fetchId: Int by remember { mutableStateOf(0) }

  // state production invocations
  ...

   // return produced state
  return Model(
    loading = currentBreed == null,
    breeds = breeds,
    currentBreed = currentBreed,
    currentUrl = currentUrl,
  )
}
```

Reply:
Passing it in the constructor works, thanks about that!
And yeah I totally see your comment about the factory approach where it's not obvious that it'll create a completely new instance of the class when it goes from cold to hot again.
I just wanted to try to mitigate the need to have to remember on the ViewModel impl, which might be easy to forget since you'd have to do it each time.

I suppose you could also do
```
class PupperPicsViewModel : MoleculeViewModel<Event, Model>(
  initialState = Model.InitialState,
) {
  private val presenter = PupperPicsPresenter(PupperPicsService())

  @Composable
  override fun models(seed: Model, events: Flow<Event>): Model {
    return presenter.present(seed, events)
  }
}
```
instead of:
```
class PupperPicsViewModel : MoleculeViewModel<Event, Model>(
  initialState = Model.InitialState,
) {
  @Composable
  override fun models(seed: Model, events: Flow<Event>): Model {
    val presenter = remember { PupperPicsPresenter(PupperPicsService()) }
    return presenter.present(seed, events)
  }
}
```
So it does stay in memory and not recreated as it was before.

The reason I was looking for an interface for the `MoleculePresenter` instead of a method, since you get the signature for free, you just do `class Foo : MoleculePresenter<Event, Model>` and you get the signature that you need instantly, don't need to remember to pass the Flow<Event> or forget to pass the `seed: Model`.

But at the same time you do get them in the `models` function in your VM, so it may be redundant too, not sure about that tradeoff 🤷‍♂️

Reply:
I suppose at least for the sample here, the simpler solution should be chosen, meaning no MoleculePresenter interface but just a static function

Reply:
Oh oops, for the past ~50 minutes I was testing things out and didn't see your last messages 🙈
But yeah you got it, that's perfect I think, and the lambda accepting the seed and the events flow feels so obvious in hindsight, can't believe I didn't think about it :D 

I still feel like passing the lambda or an interface is probably a decision for each team to make, but the more I see it the more I'd also lean towards your approach too.

Do you have access to push on top of my branch here? Maybe you can make a PR merging into mine here instead? I like your approach I think we can have that in the final PR for the molecule maintainers to also take a look at.

Reply:
This PR has now been updated with the changes discussed here 👍